### PR TITLE
FISH-796 Fix clustered singleton bugs and added test

### PR DESCRIPTION
**Pull request by lprimak**
## Description
As I upgraded the code to Hazelcast 4, one of my tests was clustered singleton.
In the process of doing this, I found a few bugs, and re-added test

## Important Info
Relates to https://github.com/payara/Payara/pull/5014 and https://github.com/payara/Payara/pull/5013

### New tests
added back tests in Payara root to test clustered singleton

### Testing Performed
automated and manual tests, as well as suite

### Testing Environment
JDK 8, Mac

## Notes for Reviewers
The PR seems big, but it's mostly new test files added

### CODE REVIEW

**Comment by lprimak** [comment]
jenkins test

**Comment by lprimak** [comment]
jenkins test

**Comment by lprimak** [comment]
jenkins test

**Comment by lprimak** [comment]
jenkins test

**Comment by lprimak** [comment]
jenkins test

**Comment by lprimak** [comment]
jenkins test

**Comment by lprimak** [comment]
jenkins test

**Comment by lprimak** [comment]
jenkins test

**Comment by dmatej** [comment]
I'm curious if this would help my current issue with Hazelcast vs. docker node start, where is some race condition blocking the docker node instance start for one minute, because hazelcast on das starts some migration before node finishes it's startup -> it fails again and again each 5 seconds, after one minute it gives up and node finishes startup. Whole test finishes in cca 135 seconds.
This situation is 5 times more often than finishing test in cca 68 seconds where is no blocking.

**Comment by lprimak** [comment]
Not sure, but my gut says no, but Hazelcast 4.1 migration *could*

File: appserver/payara-appserver-modules/payara-micro-cdi/src/main/java/fish/payara/micro/cdi/extension/cluster/ClusterScopedInterceptor.java [hunk 1]
```diff
@@ -105,11 +104,11 @@ Object postConstruct(InvocationContext invocationContext) throws Exception {
     Object preDestroy(InvocationContext invocationContext) throws Exception {
         Class<?> beanClass = invocationContext.getTarget().getClass().getSuperclass();
         Clustered clusteredAnnotation = getAnnotation(beanManager, beanClass);
-        clusteredLookup.setClusteredSessionKey(beanClass);
+        clusteredLookup.setClusteredSessionKeyIfNotSet(beanClass, clusteredAnnotation);
         IAtomicLong count = clusteredLookup.getClusteredUsageCount();
         if (count.decrementAndGet() <= 0) {
             clusteredLookup.getClusteredSingletonMap().delete(clusteredLookup.getClusteredSessionKey());
-            count.destroy();
+            count.set(0);
```
Comment (lprimak) [inline appserver/payara-appserver-modules/payara-micro-cdi/src/main/java/fish/payara/micro/cdi/extension/cluster/ClusterScopedInterceptor.java:111]: I'm not quite sure why I changed this, needs a bit further investigating

Comment (lprimak) [reply appserver/payara-appserver-modules/payara-micro-cdi/src/main/java/fish/payara/micro/cdi/extension/cluster/ClusterScopedInterceptor.java:111]: Will try to test without that change and if it passes, will revert this change, then we are good to go

File: appserver/ejb/ejb-container/src/main/java/com/sun/ejb/containers/AbstractSingletonContainer.java [hunk 2]
```diff
@@ -713,7 +713,7 @@ public void destroy(Object obj) {
                 IAtomicLong count = clusteredLookup.getClusteredUsageCount();
                 if (count.decrementAndGet() <= 0) {
                     clusteredLookup.getClusteredSingletonMap().delete(clusteredLookup.getClusteredSessionKey());
-                    count.destroy();
+                    count.set(0);
```
Comment (lprimak) [inline appserver/ejb/ejb-container/src/main/java/com/sun/ejb/containers/AbstractSingletonContainer.java:716]: same here

File: appserver/payara-appserver-modules/payara-micro-cdi/src/main/java/fish/payara/micro/cdi/extension/cluster/ClusterScopedInterceptor.java (hunk 1 above)
Comment (lprimak) [reply appserver/payara-appserver-modules/payara-micro-cdi/src/main/java/fish/payara/micro/cdi/extension/cluster/ClusterScopedInterceptor.java]: passed

**Comment by lprimak** [comment]
jenkins test

**Comment by lprimak** [comment]
jenkins test

**Comment by lprimak** [comment]
jenkins test

**Comment by lprimak** [comment]
jenkins test

**Comment by lprimak** [comment]
jenkins test

File: appserver/tests/payara-samples/samples/clustered-singleton/clustered-singleton-test/licenseheader.txt [hunk 3]
```diff
@@ -0,0 +1,43 @@
+<#if licenseFirst??>
```
Comment (MattGill98) [inline appserver/tests/payara-samples/samples/clustered-singleton/clustered-singleton-test/licenseheader.txt:1]: This looks like a leaked file from NetBeans, probably needs deleting

File: appserver/tests/payara-samples/samples/clustered-singleton/clustered-singleton-test/src/main/java/com/flowlogix/clust/singleton/interceptor/ClusteredInterceptor.java [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+ * elects to include this software in this distribution under the [CDDL or GPL
+ * Version 2] license."  If you don't indicate a single choice of license, a
+ * recipient has the option to distribute your version of this file under
+ * either the CDDL, the GPL Version 2 or to extend the choice of license to
+ * its licensees as provided above.  However, if you add GPL Version 2 code
+ * and therefore, elected the GPL Version 2 license, then the option applies
+ * only if the new code is made subject to such option by the copyright
+ * holder.
+ */
+package com.flowlogix.clust.singleton.interceptor;
```
Comment (MattGill98) [inline appserver/tests/payara-samples/samples/clustered-singleton/clustered-singleton-test/src/main/java/com/flowlogix/clust/singleton/interceptor/ClusteredInterceptor.java:40]: I'd like the packages changing to `fish.payara`, inkeeping with the other tests

File: appserver/tests/payara-samples/samples/clustered-singleton/clustered-singleton-test/src/main/java/com/flowlogix/clust/singleton/interceptor/ClusteredInterceptor.java [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+
+import com.flowlogix.clust.singleton.ClusteredSingletonInterceptedEJB;
+import java.io.Serializable;
+import javax.interceptor.AroundConstruct;
+import javax.interceptor.AroundInvoke;
+import javax.interceptor.AroundTimeout;
+import javax.interceptor.InvocationContext;
+import lombok.Cleanup;
```
Comment (MattGill98) [inline appserver/tests/payara-samples/samples/clustered-singleton/clustered-singleton-test/src/main/java/com/flowlogix/clust/singleton/interceptor/ClusteredInterceptor.java:48]: Oh god, didn't we do a tonne of work a while ago to remove Lombok?

I'm not concerned with what our answer is, but I'd like it resolved before we continue this loop of adding and removing lombok

**Review by MattGill98 — CHANGES_REQUESTED** [review]
Looks broadly good, just a couple of questions

File: appserver/tests/payara-samples/samples/clustered-singleton/clustered-singleton-test/licenseheader.txt (hunk 3 above)
Comment (lprimak) [reply appserver/tests/payara-samples/samples/clustered-singleton/clustered-singleton-test/licenseheader.txt:1]: ok

File: appserver/tests/payara-samples/samples/clustered-singleton/clustered-singleton-test/src/main/java/com/flowlogix/clust/singleton/interceptor/ClusteredInterceptor.java (hunk 4 above)
Comment (lprimak) [reply appserver/tests/payara-samples/samples/clustered-singleton/clustered-singleton-test/src/main/java/com/flowlogix/clust/singleton/interceptor/ClusteredInterceptor.java:40]: ok

File: appserver/tests/payara-samples/samples/clustered-singleton/clustered-singleton-test/src/main/java/com/flowlogix/clust/singleton/interceptor/ClusteredInterceptor.java (hunk 5 above)
Comment (lprimak) [reply appserver/tests/payara-samples/samples/clustered-singleton/clustered-singleton-test/src/main/java/com/flowlogix/clust/singleton/interceptor/ClusteredInterceptor.java:48]: ok

Comment (lprimak) [reply appserver/tests/payara-samples/samples/clustered-singleton/clustered-singleton-test/src/main/java/com/flowlogix/clust/singleton/interceptor/ClusteredInterceptor.java:48]: haha that tonne of work introduced most of these bugs :)

Comment (lprimak) [reply appserver/tests/payara-samples/samples/clustered-singleton/clustered-singleton-test/src/main/java/com/flowlogix/clust/singleton/interceptor/ClusteredInterceptor.java:48]: This time Im going to do this correctly by using delombok plugin :)

**Comment by lprimak** [comment]
jenkins test

File: appserver/tests/payara-samples/samples/clustered-singleton/clustered-singleton-test/src/main/java/fish/payara/samples/clustered/singleton/ClusteredAnnotatedAPI2.java [hunk 6]
```diff
@@ -0,0 +1,62 @@
+/*
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS HEADER.
+ *
+ * Copyright (c) [2016-2017] Payara Foundation and/or its affiliates. All rights reserved.
```
Comment (MattGill98) [inline appserver/tests/payara-samples/samples/clustered-singleton/clustered-singleton-test/src/main/java/fish/payara/samples/clustered/singleton/ClusteredAnnotatedAPI2.java:4]: Looks like there are some copyright headers that need fixing, then this is good to go IMO

**Review by MattGill98 — CHANGES_REQUESTED** [review]

**Comment by lprimak** [comment]
jenkins test

**Review by Pandrex247 — APPROVED** [review]
I haven't gone through it with anything like a fine-toothed comb, but the tests pass and no obvious purple elephants from a skim over the code.

**Review by MattGill98 — APPROVED** [review]

**Comment by AndrewG10i** [comment]
Hello there! @lprimak may I kindly share with you some early feedback as per this fix.

Yesterday I have downloaded the new release of Payara Micro 5.2021.1 (which includes this fix) and started with testing persistent schedules functionality (we use both ways of creating them with `@Schedule` and via `TimerService`), as from the fix topic it seems that persistent timers are going to be affected.

Please see below my findings considering the Payara Micro Community documentation for persistent timers -> [How EJB timers are persisted](https://docs.payara.fish/community/docs/5.2021.1/documentation/payara-micro/persistent-ejb-timers.html#how-ejb-timers-are-persisted).

In my tests we are considering two scenarios (by using cluster with two nodes launched with arguments: `--name node1` and `--name node2`):

 **Scenario 1: instance (node1) that created persistent timer goes DOWN.**

In pre-5.2021.1 versions (e.g. 5.2020.7):
The following statement from the documentation:
"If that instance goes down, the timer will be recreated on another instance with the same name once it joins the cluster."
actually worked as:
"If that instance goes down, the timer will be recreated _(running) on another instance in the cluster._"

Since 5.2021.1 seems that was fixed, and now persistent timer doesn't run in the cluster (what basically corresponds to the documentation "Until that time, the timer becomes inactive.").
Just to note here that actually my tests shown that persistent timer runs once on the another node (node2) in the cluster, but after that single run - node2 shows message like:
`[2021-02-11T13:16:34.957+0800] [] [INFO] [] [javax.enterprise.system.container.ejb.fish.payara.ejb.timer.hazelcast] The timer (1@@1613020572690@@node1@@clusterDev) is now owned by (node1). Removing from local cache`

Seems all works in compliance with the documentation (spec) now, except that timer managed to run once on another node, but okay, in general - no action needed. 

**Scenario 2: instance (node1) that created persistent timer goes UP after it was down.**

In pre-5.2021.1 versions (e.g. 5.2020.7) persistent timer was keep running on the another node (node2).

As per 5.2021.1: According to the documentation the following should happen: "If that instance goes down, **the timer will be recreated on another instance with the same name once it joins the cluster.**"
But in my tests once I launch again node1 (which has originally created the timer) - timer is not re-created on it. Basically I do not see timer running at all in the cluster anymore.

This is our **preliminary observations** on the persistent timers behavior in Payara Micro 5.2021.1 with the fix according to this ticket.
I would like to emphasize that currently we are not registering separate issue for this as we are still working on checking and reconfirming our observations, and also adapting our application to the new behavior of persistent timers (as we need to migrate to the new release due to fixes for the other bugs).

We will do our best to prepare and publish test applications and test scenarios for demonstrating the reported behavior and once it is done and fully confirmed in our tests - we will register it as a separate issue.

Thank you!

**Pull request merged**